### ISSUE TITLE
Webservice response 8.0.4 HTTP XML response is not parsable

### ISSUE THREAD

**Issue by flanderboy** [PrestaShop/PrestaShop#32476]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Hello, I'm have a new installation of 8.0.4 and i try to using webservice, if I call, for example:

```
$webService = new PrestaShopWebservice(getenv('WS_PS_API_URL'), getenv('WS_PS_KEY'), true);
$xml = $webService->get([
      'resource' => 'categories',
      'schema' => 'blank'
]);
```

the respons is 

```
Other error: <br />HTTP XML response is not parsable: array (
  0 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 68,
     'column' => 3,
     'message' => 'StartTag: invalid element name
',
     'file' => '',
     'line' => 22,
  )),
  1 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 73,
     'column' => 19,
     'message' => 'expected \'>\'
',
     'file' => '',
     'line' => 22,
  )),
  2 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 19,
     'message' => 'Opening and ending tag mismatch: category line 21 and unparseable
',
     'file' => '',
     'line' => 22,
  )),
  3 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 20,
     'message' => 'Opening and ending tag mismatch: categories line 20 and category
',
     'file' => '',
     'line' => 23,
  )),
  4 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 14,
     'message' => 'Opening and ending tag mismatch: associations line 19 and categories
',
     'file' => '',
     'line' => 24,
  )),
  5 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 68,
     'column' => 3,
     'message' => 'StartTag: invalid element name
',
     'file' => '',
     'line' => 27,
  )),
  6 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 73,
     'column' => 19,
     'message' => 'expected \'>\'
',
     'file' => '',
     'line' => 27,
  )),
  7 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 19,
     'message' => 'Opening and ending tag mismatch: product line 26 and unparseable
',
     'file' => '',
     'line' => 27,
  )),
  8 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 19,
     'message' => 'Opening and ending tag mismatch: products line 25 and product
',
     'file' => '',
     'line' => 28,
  )),
  9 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 12,
     'message' => 'Opening and ending tag mismatch: category line 3 and products
',
     'file' => '',
     'line' => 29,
  )),
  10 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 16,
     'message' => 'Opening and ending tag mismatch: prestashop line 2 and associations
',
     'file' => '',
     'line' => 30,
  )),
  11 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 5,
     'column' => 1,
     'message' => 'Extra content at the end of the document
',
     'file' => '',
     'line' => 31,
  )),
)
```

But if I try to other resource the respons is correct 

Can someone help me?
Thank you

PHP Version php8.1-fpm

### Steps to reproduce

1) Active webservice
2) Get API Key
3) try call it

### PrestaShop version(s) where the bug happened

8.0.4

### PHP version(s) where the bug happened

php8.1-fpm

**Comment by flanderboy** [comment]
no one?

**Comment by florine2623** [comment]
Hello @flanderboy ,

I can't reproduce your issue.
On my end, I'm trying to get the categories.

Please check my screen record as it is working correctly :

https://github.com/PrestaShop/PrestaShop/assets/16019289/be4f31ef-1e20-44db-aa7a-68c66db4605e

**Comment by flanderboy** [comment]
Hello @florine2623,

thanks for your answer, my problem is when a try to create a new category via PHP PrestaShopWebservice method, 
if i try to get all or one I do not receive any error with same method


Stefano

**Comment by flanderboy** [comment]
I forgot to mention that the same call in Prestashop 1.7.8.9 works perfectly

**Comment by florine2623** [comment]
ping @PrestaShop/committers ,
Could someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by gitbjorn** [comment]
We have exactly the same problem at both a dev server and at customer hosting. All we did is a clean install of Prestashop 8.0.4 on PHP 8.0.28, add a webservice key and execute (from PHP 7.4) the following request:

**HTTP REQUEST HEADER**
GET /prestashop8/api/categories?&ws_key=*****&schema=blank HTTP/2
Host: *****
User-Agent: Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US; rv:1.9.1.2) Gecko/20090729 Firefox/3.5.2 GTB5
Accept: */*

**HTTP RESPONSE HEADER**
HTTP/2 200 
server: nginx
date: Mon, 22 May 2023 05:51:09 GMT
content-type: text/xml;charset=utf-8
access-time: 1684734669
x-powered-by: PrestaShop Webservice
psws-version: 8.0.4
execution-time: 0.008
content-sha1: 90530ddbe34d3fa1e1534d5d7ccb170e3ab65a5a
vary: Accept-Encoding
x-powered-by: PHP

**RETURN HTTP BODY**
```
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
<category>
	<id><![CDATA[]]></id>
	<id_parent><![CDATA[]]></id_parent>
	<active><![CDATA[]]></active>
	<id_shop_default><![CDATA[]]></id_shop_default>
	<is_root_category><![CDATA[]]></is_root_category>
	<position><![CDATA[]]></position>
	<date_add><![CDATA[]]></date_add>
	<date_upd><![CDATA[]]></date_upd>
	<name><language id="1"><![CDATA[]]></language></name>
	<link_rewrite><language id="1"><![CDATA[]]></language></link_rewrite>
	<description><language id="1"><![CDATA[]]></language></description>
	<additional_description><language id="1"><![CDATA[]]></language></additional_description>
	<meta_title><language id="1"><![CDATA[]]></language></meta_title>
	<meta_description><language id="1"><![CDATA[]]></language></meta_description>
	<meta_keywords><language id="1"><![CDATA[]]></language></meta_keywords>
<associations>
<categories>
	<category>
	<0><![CDATA[]]></0>
	</category>
</categories>
<products>
	<product>
	<0><![CDATA[]]></0>
	</product>
</products>
</associations>
</category>
</prestashop>
```

**[ERROR]** 
HTTP XML response is not parsable: array (
  0 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 68,
     'column' => 3,
     'message' => 'StartTag: invalid element name
',
     'file' => '',
     'line' => 22,
  )),
  1 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 73,
     'column' => 19,
     'message' => 'expected \'>\'
',
     'file' => '',
     'line' => 22,
  )),
  2 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 19,
     'message' => 'Opening and ending tag mismatch: category line 21 and unparsable
',
     'file' => '',
     'line' => 22,
  )),
  3 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 13,
     'message' => 'Opening and ending tag mismatch: categories line 20 and category
',
     'file' => '',
     'line' => 23,
  )),
  4 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 14,
     'message' => 'Opening and ending tag mismatch: associations line 19 and categories
',
     'file' => '',
     'line' => 24,
  )),
  5 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 68,
     'column' => 3,
     'message' => 'StartTag: invalid element name
',
     'file' => '',
     'line' => 27,
  )),
  6 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 73,
     'column' => 19,
     'message' => 'expected \'>\'
',
     'file' => '',
     'line' => 27,
  )),
  7 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 19,
     'message' => 'Opening and ending tag mismatch: product line 26 and unparsable
',
     'file' => '',
     'line' => 27,
  )),
  8 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 12,
     'message' => 'Opening and ending tag mismatch: products line 25 and product
',
     'file' => '',
     'line' => 28,
  )),
  9 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 12,
     'message' => 'Opening and ending tag mismatch: category line 3 and products
',
     'file' => '',
     'line' => 29,
  )),
  10 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 76,
     'column' => 16,
     'message' => 'Opening and ending tag mismatch: prestashop line 2 and associations
',
     'file' => '',
     'line' => 30,
  )),
  11 => 
  LibXMLError::__set_state(array(
     'level' => 3,
     'code' => 5,
     'column' => 1,
     'message' => 'Extra content at the end of the document
',
     'file' => '',
     'line' => 31,
  )),
)

I guess the error is in

    <0><![CDATA[]]></0>

**Comment by gitbjorn** [comment]
This issue doesn't occur on PHP 8.2

**Comment by gitbjorn** [comment]
@florine2623 

> Hello @flanderboy ,
> 
> I can't reproduce your issue. On my end, I'm trying to get the categories.
> 
> Please check my screen record as it is working correctly :
> 
>  Screen.Recording.2023-05-09.at.16.08.20.mov

Can you try and get the categories with ?schema=blank ? The issue only seems to be in the categories, for example products?schema=blank works

**Comment by gitbjorn** [comment]
Besides the categories?schema=blank, this issue also occurs for combinations?schema=blank . I have tested on PHP 8.0, 8.1 and 8.2. 

Also, I read somewhere that this issue would be resolved in 8.1, however, the 8.1 beta of February 25th still has this issue

**Comment by ghost** [comment]
To map out the problem in more detail, you can perform the following simple checks.
When connecting to the Prestashop Webservice API, you use the following settings:

Make sure your PrestaShop web service is **_enabled_** and create a api key with at least the permissions resource _**combinations**_ selected.

Accessing directly trough the browser using url **[https://YOUR_API_KEY@YOUR_STORE_NAME/api/combinations?schema=blank](https://YOUR_API_KEY@YOUR_STORE_NAME/api/combinations?schema=blank)**
For local-host we can use **http://YOUR_API_KEY@127.0.0.1/YOUR_STORE_NAME/api/combinations?schema=blank**

**Prestashop 1.7.8.9** 
PHP versions: 7.4.33
Response:
```
<prestashop>
<combination>
<id></id>
<id_product></id_product>
<location></location>
<ean13></ean13>
<isbn></isbn>
<upc></upc>
<mpn></mpn>
<quantity></quantity>
<reference></reference>
<supplier_reference></supplier_reference>
<wholesale_price></wholesale_price>
<price></price>
<ecotax></ecotax>
<weight></weight>
<unit_price_impact></unit_price_impact>
<minimal_quantity></minimal_quantity>
<low_stock_threshold></low_stock_threshold>
<low_stock_alert></low_stock_alert>
<default_on></default_on>
<available_date></available_date>
<associations>
</associations>
</combination>
</prestashop>
``` 

----------------------------------------------------------------------------------------------------------------

**Prestashop 1.7.8.9**
PHP versions: 8.0.26, 8.1.13, 8.2.0
Response:
``` 
XML parse error: malformed
Location: http://127.0.0.1/devstore/api/combinations?schema=blank
Rule number 27, column 3:
	<0><![CDATA[]]></0>
---------^
``` 

----------------------------------------------------------------------------------------------------------------

**Prestashop 1.8.0.4**
PHP versions: 7.4.33, 8.0.26, 8.1.13, 8.2.0
Response:
``` 
XML parse error: malformed
Location: http://127.0.0.1/devstore_8/api/combinations?schema=blank
Rule number 27, column 3:
	<0><![CDATA[]]></0>
---------^
``` 

Using the Chrome browser we get a different error 
```
This page contains the following errors:
error on line 27 at column 3: StartTag: invalid element name
Below is a rendering of the page up to the first error.
```

When we debug the connection it self we receive following error:
```
HTTP XML response is not parsable:
array(
    0 => LibXMLError::__set_state(array('level' => 3, 'code' => 68, 'column' => 3, 'message' => 'StartTag: invalid element name ', 'file' => '', 'line' => 25,)),
    1 => LibXMLError::__set_state(array('level' => 3, 'code' => 73, 'column' => 19, 'message' => 'expected \'>\' ', 'file' => '', 'line' => 25,)),
    2 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 19, 'message' => 'Opening and ending tag mismatch: product_option_value line 24 and unparseable ', 'file' => '', 'line' => 25,)),
    3 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 45, 'message' => 'Opening and ending tag mismatch: product_option_values line 24 and product_option_value ', 'file' => '', 'line' => 26,)),
    4 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 25, 'message' => 'Opening and ending tag mismatch: associations line 24 and product_option_values ', 'file' => '', 'line' => 27,)),
    5 => LibXMLError::__set_state(array('level' => 3, 'code' => 68, 'column' => 3, 'message' => 'StartTag: invalid element name ', 'file' => '', 'line' => 30,)),
    6 => LibXMLError::__set_state(array('level' => 3, 'code' => 73, 'column' => 19, 'message' => 'expected \'>\' ', 'file' => '', 'line' => 30,)),
    7 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 19, 'message' => 'Opening and ending tag mismatch: image line 24 and unparseable ', 'file' => '', 'line' => 30,)),
    8 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 15, 'message' => 'Opening and ending tag mismatch: images line 24 and image ', 'file' => '', 'line' => 31,)),
    9 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 10, 'message' => 'Opening and ending tag mismatch: combination line 24 and images ', 'file' => '', 'line' => 32,)),
    10 => LibXMLError::__set_state(array('level' => 3, 'code' => 76, 'column' => 16, 'message' => 'Opening and ending tag mismatch: prestashop line 24 and associations ', 'file' => '', 'line' => 33,)),
    11 => LibXMLError::__set_state(array('level' => 3, 'code' => 5, 'column' => 1, 'message' => 'Extra content at the end of the document ', 'file' => '', 'line' => 34,)),
);
```

When we use JSON instead of XML (DevDocs [https://devdocs.prestashop-project.org/8/webservice/getting-started/#using-json-instead-of-xml](https://devdocs.prestashop-project.org/8/webservice/getting-started/#using-json-instead-of-xml)

We receive following result in **Prestashop 1.7.8.9**
```
{
    "combination": {
        "id":"",
        "id_product":"",
        "location":"",
        "ean13":"",
        "isbn":"",
        "upc":"",
        "mpn":"",
        "quantity":"",
        "reference":"",
        "supplier_reference":"",
        "wholesale_price":"",
        "price":"",
        "ecotax":"",
        "weight":"",
        "unit_price_impact":"",
        "minimal_quantity":"",
        "low_stock_threshold":"",
        "low_stock_alert":"",
        "default_on":"",
        "available_date":"",
        "associations": {
            "product_option_values":[[null]], 
            "images":[[null]]
        }
    }
}
```

And following result in **Prestashop 1.8.0.4**
```
{
    "combination": {
        "associations": {
            "product_option_values":[[null]], 
            "images":[[null]]
        }
    }
}
```

**Comment by ghost** [comment]
I received comments regarding a problem that was supposed to be solved in version 8.1. However, after investigating, I could only find a pull request at this link: [GitHub Pull Request](https://github.com/PrestaShop/PrestaShop/pull/31023/files), and it does not seem to address the problem adequately.

Additionally, we examined the XML format requirement, which states that XML elements must not start with a number. This violates proper programming conventions.

Currently, the problem only occurs with product combinations and categories.

To resolve this, follow these steps:

1. Open the file located at `YOUR_STORE/classes/webservice/WebserviceOutputBuilder.php`.
2. Navigate to line 718, where you will find the following code snippet:

```php
} elseif (!isset($field['sqlId'])) {
    $field['sqlId'] = $field_name;
    $field['value'] = isset($object_assoc[$field_name]) ? $object_assoc[$field_name] : null;
}
``` 

3. Replace the line `$field['sqlId'] = $field_name;` with the following line:

```php
$field['sqlId'] = $field_name ?: 'id';
``` 

By using the PHP shorthand version (?:) of if/else, we check if **$field_name** has a value. If it does not, we assign '**id**' as the default value. This ensures that the start tag does not have the value "0," thereby preventing the generation of an error message.

**Comment by matks** [comment]
Hello everybody I was able to reproduce the issue 👍 I confirm the bug
